### DISCUSSION TITLE
styling issue with Webchat due to the font-size being set to 62.5%

### DISCUSSION THREAD

**Discussion by sakukreti** [twilio-labs/paste#3064] [General] [answered]
Hi team, I have a customer who is encountering a styling issue with Webchat due to the font-size being set to 62.5%  on one of the pages the Webchat-Widget is included in.
They are using Twilio-Paste Components and cannot change the font-size to 100% as that would break the rest of a page. Do you have a solution to accommodate for that or documentation that would help us fix it? 

PS: Setting font-size to 62.5% in the html tag is a popular trick to ease the styling using rem values: https://www.toheeb.com/en/pages/62.5-font-size-trick/
They are using https://github.com/twilio/twilio-webchat-react-app/ as base repo 

To give some visual context here...When they open the chat bubble and the webchat app is visible, the page change the size of its components. If they open the bubble when customer is logged in (another page), in this case the page is ok but the chat elements are too little. Sometimes it's changing the page style and some others the webchat ones. Must be something in between that is causing this issue
![image](https://user-images.githubusercontent.com/118907937/222357500-a9e1ac69-1bc4-4179-bff0-becb012b9b71.png)
![image](https://user-images.githubusercontent.com/118907937/222357588-47b0faaf-066f-4eb1-a1e0-113315bee7de.png)
![image](https://user-images.githubusercontent.com/118907937/222357657-56fc0610-13ce-4dca-bf24-cdb20b800d05.png)

**Comment by nkrantz** [comment]
Hi @sakukreti - we're taking a look at this today and we'll get back to you soon with an update. Thanks!

**Comment by SiTaggart** [comment]
Hi @sakukreti 

To be able to correctly set proportionally sized spacing, sizing and typography based on the users personal preference, we have to rely on a percentage font-size set on the html element. We do this for accessibility reasons. 

There are two schools of thought, those who set the base font size to 100% (16px) and those who set the base font size to 62.5% (10px). Neither are wrong, they work in exactly the same way. But we had to pick one, and we predominantly saw 100% being set on the applications we were targeting. Unfortunately, that doesn't work in this case.

I think ultimately the only thing we can do is add a new design token for base font size. A consumer could then override the base font size token, and font-size tokens to match their application's needs, so Paste doesn't clash.

I can't say when we'd get to that, though, at the moment, as we're in the midst of closing out Q1 right now. We'll come back and update when we've gotten it triaged.

**Comment by espipj** [comment]
Hello related with this issue. We've been investigating on the webchat ui side and it seems that Paste is adding styles on the body itself whenever you add a styles provider in your app (in this case the webchat). One of those styles is related to the fonts sizing. is this expected? If so, is there any workaround to override those global styles/remove them customer side?
![image](https://user-images.githubusercontent.com/12865914/223121227-d03e1399-8769-47ae-84ef-668ea2dd4209.png)

**Reply by SiTaggart** [reply]
Yeah, that one is solvable with tokens today, which I briefly hinted at in my original response. 

Every font size is just a font size token. For something like embedded web chat where you need to live in someone else site, you'd probably need to convert all font size tokens to pixel values.

From memory the base size is `FontSize30` for us, which is what we set on the body. If that is set to a pixel value that matches the host website style, and you create a type scale based on that, you should get a lot of success with the Customization Provider.

100% on the HTML is another problem, that a new token would solve, in the same way

**Reply by oscar8880** [reply]
Hey both, I've mocked up this quick sandbox so we can see the effect of overriding fontSize30 with the main webpage's base font size: https://codesandbox.io/s/paste-html-font-size-workaround-kh1kyp?file=/src/index.tsx

@SiTaggart the theory about setting that pixel value is correct and we can see that it maintains the main webpage font sizes at their normal size. However there is a side effect then on the Webchat app since its default font size is determined by fontSize30, e.g. here where I've set it to 10px for illustration.
![Screenshot 2023-03-07 at 11 45 44](https://user-images.githubusercontent.com/33548913/223415404-06e8eb61-b6a8-4603-97db-c0f3109e14c9.png)

If the customer's base font size is similar to Webchat app's then the difference may be negligible, so I think we can still suggest this as a temporary workaround for them. Otherwise as mentioned in your original answer the new token to control the base font size is our route to a lasting solution.

cc @sakukreti

**Reply by sakukreti** [reply]
Thanks @oscar8880 the customer worked around the styling issue by creating their own custom theme that's based on a font-size of 62.5%. One issue they are currently struggling to resolve is that on some pages, where the font-size is set at 62.5%, opening the webchat widget, which is made of Twilio-paste components, sets the font-size to 100%. 
This destroys the styling for the rest of the page. They could pinpoint the HTML element to:
<style data-emotion="css-global" data-s=""></style>
Removing the HTML element fixes the styling.

**Reply by oscar8880** [reply]
No problem @sakukreti

The root cause is Paste applying the font-size 100% style to `html` global styles: https://github.com/twilio-labs/paste/blob/a4c9e70b0820bfb2a41f58a61e7f991d9fa3a09b/packages/paste-theme/src/styles/global.ts#L5 which adds that `style` tag you mention.

As @SiTaggart mentioned in his [original answer](https://github.com/twilio-labs/paste/discussions/3064#discussioncomment-5198169) a long term solution is a Paste token which allows that `html { font-size }` setting to be overridden, so the customer could set that to 62.5% to match their own styles.

As a workaround the customer can try setting the fontSize30 token to the pixel value of their base font size. That should fix the issue with their font styles being changed when Webchat is rendered, with the potential side effect of a change in the font on Webchat.

**Reply by sakukreti** [reply]
@oscar8880 the customer tried the workaround but it did not work, they suspect we dont see the same issue in Codesanbox because it does not have any <style data-emotion="css-global" data-s=""></style> tag included.

**Reply by oscar8880** [reply]
@sakukreti 

Checking the source of the code sandbox I do see the `<style data-emotion="css-global" data-s=""></style>` tag:
<img width="923" alt="Screenshot 2023-03-07 at 14 36 40" src="https://user-images.githubusercontent.com/33548913/223456217-7ba3dd38-3c86-40f8-9248-3af2b7b379de.png">. I'm not convinced that this is necessarily related to the issue.

Could you share how the customer is overriding the fontSize30 token so we can verify that they're doing that correctly?

I would expect them to be adding the override either:
- directly where we use the CustomizationProvider in the code in `src/components/WebchatWidget.tsx`, e.g on line 35:
```theme={{ ...theme?.overrides, fontSizes: { fontSize30: "16px" } }}```
- or in config which is passed to the initWebchat function in `public/index.html`:
```
Twilio.initWebchat({
	serverUrl: "%REACT_APP_SERVER_URL%",
	theme: {
		isLight: true,
		overrides : {
			fontSizes: {
				fontSize30: "16px"
			}
		}
	}
```

**Reply by SiTaggart** [reply]
We might need both, the new token for setting the `html` element font size and move setting the base font size from the `body` element back to the `Theme.Provider` and `CustomizationProvider` `div` (where it used to be a couple of years ago).

It's the closest to style encapsulation as we can get without a shadow root or iframe, given CSS is close and inherits.

**Reply by oscar8880** [reply]
That's true! Is there a way we can track progress on those two pieces of work?

**Reply by SiTaggart** [reply]
Hi folks, this is now released

**Comment by sakukreti** [comment]
@oscar8880 @SiTaggart The customer  tried all our suggestions, however, they didn't work/were not enough. 
In the end they created a custom theme that overrides all Twilio-Paste styling fields that use rem, with a new rem value, that fits to their 62.5% base (e.g. 1 rem became 1.6rem).
To stop webchat from increasing the base font-size for the whole page, they contacted the host page and asked them to add a '!important' to their font-size. For now, it seems like they managed to work around all Twilio-Paste related issues.

**Comment by SiTaggart** [comment, marked as answer]
Hi, you can follow along on progress here https://github.com/twilio-labs/paste/pull/3090

**Reply by SiTaggart** [reply]
This is now released and available to install with the latest version of Paste Core.

**Comment by sakukreti** [comment]
@SiTaggart The customer tried out the latest Paste package and do not see any results.

Steps taken:
1. Update Twilio Paste packages to latest versions:
a. @twilio-paste/core : 19.0.0,
b. /icons to 11.0.0
c. and /theme to 10.0.0

2. Override fontSizeBase token:
a. Exactly as suggested in discussion [here](https://github.com/twilio-labs/paste/discussions/3064#discussioncomment-5227254):
b. theme={{ ...theme?.overrides, fontSizes: { fontSizeBase: '62.5%' }}}
    i. in <CustomizationProvider>
c. fontSizeBase is properly recognised as a token after the update

3. Tried different values for the fontSizeBase:
a. 62.5%
b. 62.5% !important
c. 10px
d. 10px !important

None of the tested values yielded any results or made any difference, the styling issue persisted.

**Reply by SiTaggart** [reply]
My mistake, onece this PR is merged, then it'll be available to install. Hang in there https://github.com/twilio-labs/paste/pull/3126

**Reply by SiTaggart** [reply]
I actually mis-spoke about getting it wrong. It is in Paste 19. 

Here is a Codesandbox showing using an existing style sheet which sets the `html` element to 62.5%, and then customizing the Paste theme to honor that setting. https://codesandbox.io/s/base-font-size-overrides-hewmpd?file=/public/index.html

**Reply by sakukreti** [reply]
@SiTaggart the customer tried the latest solution (https://github.com/twilio-labs/paste/discussions/3064#discussioncomment-5513282) and it's working fine with the Codesanbox example provided but they were under the impression that they only have to change the fontSizeBase and internally Paste will handle/calculate the sizes taking into account the percentage of the font size base, that way they can get rid of the custom theme. 
But actually they have to do the same that they were doing with the custom theme, adding the fontSizeBase and update all values and are unsure if its worth the effort to change it.
Do you have any recommendations?

**Reply by SiTaggart** [reply]
Hi @sakukreti 

No, that was never going to be the case unfortunately, based on the fact the Paste uses proportional sizing for spacing and typography, which is influenced by the HTML font size.

This solution, although a little more manual, gives full control over to the implementor as we can't tell what it _should be_ as we can't see the application it's being added to.
